### PR TITLE
Add Image Scale forceRatio Option

### LINKED ISSUE #457: [MediaBundle] Add tests for ImageConverterCommands
<a href="https://github.com/danrot"></a>
By @danrot *from* https://github.com/sulu-cmf/SuluMediaBundle/issues/38

### LINKED ISSUE #809: Scale smaller images to right resolution
If the image format is 1200x600 and the image has only 1000x1000 pixel the image should be scaled to 1000x500. Currently it will not scaled and will return 1000x1000 image which will destroy the theme design.

### PR DESCRIPTION
**Pull request by alexander-schranz**
__Tasks:__

- [x] test coverage
- [ ] gather feedback for my changes
- [x] submit changes to the documentation
- [x] added changelog line

__Informations:__

| Q             | A
| ------------- | ---
| Tests pass?   | yes
| Fixed tickets | fixes #809, fixes #457 
| BC Breaks     | none
| Doc           | https://github.com/sulu-io/docs/blob/master/detail-specification/500-assets/DET-507-ImageConverter.md

### CODE REVIEW

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/Command/ScaleCommand.php [hunk 1]
```diff
@@ -22,6 +22,7 @@ public function execute(&$image, $parameters)
         $size = $image->getSize();
 
         $retina = isset($parameters['retina']) && $parameters['retina'] != 'false' ? 2 : 1;
+        $forceRatio = !isset($parameters['forceRatio']) || $parameters['forceRatio'] != 'false' ? true : false;
```
Comment (dantleech) [inline src/Sulu/Bundle/MediaBundle/Media/ImageConverter/Command/ScaleCommand.php]: No need for ternary operator here. It is already `true|false` I think :)

Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Media/ImageConverter/Command/ScaleCommand.php]: did update the xml reader that 'true' or 'false' is converted to boolean.

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/Command/ScaleCommand.php [hunk 2]
```diff
@@ -21,7 +21,8 @@ public function execute(&$image, $parameters)
     {
         $size = $image->getSize();
 
-        $retina = isset($parameters['retina']) && $parameters['retina'] != 'false' ? 2 : 1;
+        $retina = isset($parameters['retina']) && $parameters['retina'] ? 2 : 1;
+        $forceRatio = !isset($parameters['forceRatio']) ? true : $parameters['forceRatio'];
```
Comment (dantleech) [inline src/Sulu/Bundle/MediaBundle/Media/ImageConverter/Command/ScaleCommand.php]: For me this reads better:

````php
$forceRatio = isset($parameters['forceRatio']) ? $parameters['forceRatio'] : true;
````

I would also consider:

````php
$parameters = array_merge(array(
    'retina' => 1,
    'forceRatio' => true,
    'x' => null,
    'y' => null,
), $parameters);
````

That way you clearly define the default values and do not have to check if things are set.

Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Media/ImageConverter/Command/ScaleCommand.php]: thats a good solution thx

File: src/Sulu/Bundle/MediaBundle/Media/FormatLoader/XmlFormatLoader.php [hunk 3]
```diff
@@ -92,7 +92,13 @@ private function parseXml($file)
...
                     $parameterNodes = $this->xpath->query('x:parameters/x:parameter', $commandNode);
                     foreach ($parameterNodes as $parameterNode) {
-                        $parameters[$parameterNode->attributes->getNamedItem('name')->nodeValue] = $parameterNode->nodeValue;
+                        $value = $parameterNode->nodeValue;
+                        if ($value === 'true') {
+                            $value = true;
+                        } elseif ($value === 'false') {
+                            $value = false;
+                        }
+                        $parameters[$parameterNode->attributes->getNamedItem('name')->nodeValue] = $value;
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/FormatLoader/XmlFormatLoader.php]: What's with all the other values not covered by your if statement before? And if it is only allowed to be `'true'` or `'false'` something like the following would be sufficient:

```php
$value = $parameterNode->nodeValue === 'true';
```

**Comment by danrot** [comment]
Can't find any tests ;-)

**Comment by danrot** [comment]
And the option in the documentation is not very helpful...

File: src/Sulu/Bundle/MediaBundle/Media/FormatLoader/XmlFormatLoader.php (hunk 3 above)
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Media/FormatLoader/XmlFormatLoader.php]: this is for all parameters.
if its not 'true' or 'false' the value whatever it is is set as parameter value 
https://github.com/sulu-io/sulu/blob/bugfix/image-keep-resolution/src/Sulu/Bundle/MediaBundle/Media/FormatLoader/XmlFormatLoader.php#L95

**Comment by alexander-schranz** [comment]
documentation updated

File: src/Sulu/Bundle/MediaBundle/Media/ImageConverter/Command/ScaleCommand.php [hunk 4]
```diff
@@ -19,21 +19,83 @@ class ScaleCommand implements CommandInterface
      */
     public function execute(&$image, $parameters)
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/ImageConverter/Command/ScaleCommand.php]: Can you please add a type hint to `$image`? It's not very easy to see what kind of object is handled here.

**Comment by danrot** [comment]
Tests are still missing :-)

**Comment by alexander-schranz** [comment]
@danrot tests added and branch rebased

File: src/Sulu/Bundle/MediaBundle/Tests/Unit/Media/ImageConverter/Command/AbstractCommandTest.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            $image = $imagine->create($imageBox);
+
+            $this->command->execute($image, $data['options']);
+
+            $this->assertEquals($data['width'], $image->getSize()->getWidth());
+            $this->assertEquals($data['height'], $image->getSize()->getHeight());
+        }
+    }
+
+
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Tests/Unit/Media/ImageConverter/Command/AbstractCommandTest.php]: No new lines

Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Tests/Unit/Media/ImageConverter/Command/AbstractCommandTest.php]: @danrot lines removed

**Pull request merged**